pkg/orm/types.go

```diff
@@ -74,63 +82,93 @@ type Ormer interface {
 	// args are limit, offset int and order string.
 	//
 	// example:
 	// 	Ormer.LoadRelated(post,"Tags")
 	// 	for _,tag := range post.Tags{...}
-	//args[0] bool true useDefaultRelsDepth ; false  depth 0
-	//args[0] int  loadRelationDepth
-	//args[1] int limit default limit 1000
-	//args[2] int offset default offset 0
-	//args[3] string order  for example : "-Id"
+	// args[0] bool true useDefaultRelsDepth ; false  depth 0
+	// args[0] int  loadRelationDepth
+	// args[1] int limit default limit 1000
+	// args[2] int offset default offset 0
+	// args[3] string order  for example : "-Id"
 	// make sure the relation is defined in model struct tags.
-	LoadRelated(md interface{}, name string, args ...interface{}) (int64, error)
+	LoadRelated( md interface{}, name string, args ...interface{}) (int64, error)
+	LoadRelatedWithCtx(ctx context.Context, md interface{}, name string, args ...interface{}) (int64, error)
+
 	// create a models to models queryer
 	// for example:
 	// 	post := Post{Id: 4}
 	// 	m2m := Ormer.QueryM2M(&post, "Tags")
-	QueryM2M(md interface{}, name string) QueryM2Mer
+	QueryM2M( md interface{}, name string) QueryM2Mer
+	QueryM2MWithCtx(ctx context.Context, md interface{}, name string) QueryM2Mer
+
 	// return a QuerySeter for table operations.
 	// table name can be string or struct.
 	// e.g. QueryTable("user"), QueryTable(&user{}) or QueryTable((*User)(nil)),
 	QueryTable(ptrStructOrTableName interface{}) QuerySeter
+	QueryTableWithCtx(ctx context.Context, ptrStructOrTableName interface{}) QuerySeter
+
 	// switch to another registered database driver by given name.
-	Using(name string) error
-	// begin transaction
-	// for example:
-	// 	o := NewOrm()
-	// 	err := o.Begin()
-	// 	...
-	// 	err = o.Rollback()
-	Begin() error
-	// begin transaction with provided context and option
-	// the provided context is used until the transaction is committed or rolled back.
-	// if the context is canceled, the transaction will be rolled back.
-	// the provided TxOptions is optional and may be nil if defaults should be used.
-	// if a non-default isolation level is used that the driver doesn't support, an error will be returned.
-	// for example:
-	//  o := NewOrm()
-	// 	err := o.BeginTx(context.Background(), &sql.TxOptions{Isolation: sql.LevelRepeatableRead})
-	//  ...
-	//  err = o.Rollback()
-	BeginTx(ctx context.Context, opts *sql.TxOptions) error
-	// commit transaction
-	Commit() error
-	// rollback transaction
-	Rollback() error
-	// return a raw query seter for raw sql string.
-	// for example:
-	//	 ormer.Raw("UPDATE `user` SET `user_name` = ? WHERE `user_name` = ?", "slene", "testing").Exec()
-	//	// update user testing's name to slene
-	Raw(query string, args ...interface{}) RawSeter
-	Driver() Driver
+	// Using(name string) error
+
 	DBStats() *sql.DBStats
 }
 
+type DriverGetter interface {
+	Driver() Driver
+}
+
+type Ormer interface {
+	DQL
+	DML
+	DriverGetter
+	TxBeginner
+}
+
+type TxOrmer interface {
```

Comment:
I think we should allow users to start a new transaction when they use TxOrmer. So that they can execute the transaction inside another transaction.

Reply:
sql.Tx not provide method **Begin**. So i think sholud not provide transaction in TxOrmer